### PR TITLE
Fix Python interop unit-tests on Windows

### LINKED ISSUE #5163: Fix Python tests for upstream/master on Windows
Currently this is making master build red.

In terms of priorities, this should be done after we have the artifacts building on windows.

If this is not currently feasible, let's disable the tests on windows so we see green again.

### PR DESCRIPTION
**Pull request by soltanmm**
Depends on #6797, which depends on #6791.

Fixes #5163.

### CODE REVIEW

**Comment by soltanmm** [comment]
Python Windows is green on basic tests.

**Comment by soltanmm** [comment]
Zomg this got git-insensitive-rebase-conflict-sniped.

File: tools/run_tests/run_tests.py [hunk 1]
```diff
@@ -431,23 +420,50 @@ def makefile_name(self):
...
+    python_config_generator = lambda name, major, minor, bits: PythonConfig(
+        name,
+        shell + builder + builder_prefix_arguments
+	    + [python_pattern_function(major=major, minor=minor, bits=bits)]
+	    + [name] + venv_relative_python + toolchain,
+        shell + runner + [os.path.join(name, venv_relative_python[0])])
+    python27_config = python_config_generator(name='py27', major='2', minor='7', bits=bits)
+    python34_config = python_config_generator(name='py34', major='3', minor='4', bits=bits)
+    if args.compiler == 'default':
+      if os.name == 'nt':
```
Comment (soltanmm) [inline tools/run_tests/run_tests.py]: I'll note that this was a strange decision on my part. My gut justification is that Python 3 support is AFAIK in `beta`, so dragging Windows along immediately seemed a bit much (correct me if I'm wrong).

File: tools/run_tests/run_tests.py [hunk 2]
```diff
@@ -431,23 +420,50 @@ def makefile_name(self):
...
+      python_pattern_function = lambda major, minor, bits: 'python{major}.{minor}'.format(major=major, minor=minor)
+    runner = [os.path.abspath('tools/run_tests/run_python.sh')]
+    python_config_generator = lambda name, major, minor, bits: PythonConfig(
+        name,
+        shell + builder + builder_prefix_arguments
+	    + [python_pattern_function(major=major, minor=minor, bits=bits)]
+	    + [name] + venv_relative_python + toolchain,
+        shell + runner + [os.path.join(name, venv_relative_python[0])])
+    python27_config = python_config_generator(name='py27', major='2', minor='7', bits=bits)
+    python34_config = python_config_generator(name='py34', major='3', minor='4', bits=bits)
```
Comment (soltanmm) [inline tools/run_tests/run_tests.py]: @thunderboltsid If this PR goes in in this form, the `PythonConfig` from before will be a bit simplified relative to #6791.

File: src/python/grpcio/build.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
+# Copyright 2016, Google Inc.
```
Comment (kpayson64) [inline src/python/grpcio/build.py]: I think you might be able to simplify this:

If my understanding is correct, this is a workaround the windows command line character limit. What if instead of monkeypatching the entire link operation, you just monkeypatch the spawn() operation?

File: tools/run_tests/build_python.sh [hunk 4]
```diff
@@ -61,17 +127,24 @@ if [ "${PLATFORM/Darwin}" = "$PLATFORM" ]; then
...
+if [ $(is_mingw) ]; then
+  # We're on MinGW, and our CFLAGS and LDFLAGS will be eaten by the void. Use
+  # our work-around environment variables instead.
+  PYTHON_MSVCR=`$PYTHON -c "from distutils.cygwinccompiler import get_msvcr; print(get_msvcr()[0])"`
+  export GRPC_PYTHON_LDFLAGS="-static-libgcc -static-libstdc++ -mcrtdll=$PYTHON_MSVCR -static -lpthread"
+  # See https://sourceforge.net/p/mingw-w64/bugs/363/
+  export GRPC_PYTHON_CFLAGS="-D_ftime=_ftime64 -D_timeb=__timeb64"
+  # TODO(atash) set these flags for only grpcio-tools (they don't do any harm to
+  # grpcio, but they result in noisy warnings).
+  export GRPC_PYTHON_CFLAGS="-frtti -std=c++11 $GRPC_PYTHON_CFLAGS"
```
Comment (kpayson64) [inline tools/run_tests/build_python.sh]: There were some issues with the c++11 standard on Mac, can we just use the c++98 standard on all our platforms?

File: tools/run_tests/build_python.sh [hunk 5]
```diff
@@ -61,17 +127,24 @@ if [ "${PLATFORM/Darwin}" = "$PLATFORM" ]; then
...
+# TODO(atash) consider conceptualizing MinGW as a first-class platform and move
+# these flags into our `setup.py`s
+if [ $(is_mingw) ]; then
+  # We're on MinGW, and our CFLAGS and LDFLAGS will be eaten by the void. Use
+  # our work-around environment variables instead.
+  PYTHON_MSVCR=`$PYTHON -c "from distutils.cygwinccompiler import get_msvcr; print(get_msvcr()[0])"`
+  export GRPC_PYTHON_LDFLAGS="-static-libgcc -static-libstdc++ -mcrtdll=$PYTHON_MSVCR -static -lpthread"
+  # See https://sourceforge.net/p/mingw-w64/bugs/363/
+  export GRPC_PYTHON_CFLAGS="-D_ftime=_ftime64 -D_timeb=__timeb64"
+  # TODO(atash) set these flags for only grpcio-tools (they don't do any harm to
```
Comment (kpayson64) [inline tools/run_tests/build_python.sh]: It looks like these end up getting set for both grpcio and grpcio-tools.

File: tools/run_tests/run_tests.py (hunk 1 above)
Comment (kpayson64) [inline tools/run_tests/run_tests.py]: I'd like to get python2.7 for windows in before we start worrying about python 3

File: PYTHON-MANIFEST.in [hunk 6]
```diff
@@ -7,6 +7,7 @@ graft include/grpc
 graft third_party/boringssl
 graft third_party/nanopb
 graft third_party/zlib
+include src/python/grpcio/build.py
```
Comment (kpayson64) [inline PYTHON-MANIFEST.in]: I'm not a huge fan of calling this build.py, but I don't feel strongly about it.

File: tools/distrib/python/grpcio_tools/setup.py [hunk 7]
```diff
@@ -60,8 +64,9 @@
 GRPC_PYTHON_TOOLS_PACKAGE = 'grpc.tools'
 GRPC_PYTHON_PROTO_RESOURCES_NAME = '_proto'
 
-import protoc_lib_deps
-import grpc_version
+DEFINE_MACROS = (('HAVE_PTHREAD', 1),)
+if "win32" in sys.platform and '64bit' in platform.architecture()[0]:
+  DEFINE_MACROS += (('MS_WIN64', 1),)
```
Comment (kpayson64) [inline tools/distrib/python/grpcio_tools/setup.py]: Is this needed?

File: tools/run_tests/build_python.sh (hunk 4 above)
Comment (soltanmm) [inline tools/run_tests/build_python.sh]: That'd be for post-#7208. Or a modification to it.

File: tools/run_tests/build_python.sh (hunk 5 above)
Comment (soltanmm) [inline tools/run_tests/build_python.sh]: Methinks the `TODO` is a sufficient response. :-)

File: PYTHON-MANIFEST.in (hunk 6 above)
Comment (soltanmm) [inline PYTHON-MANIFEST.in]: I'd honestly love to get it renamed to something like `HACKY-HACKY-THING-BECAUSE-DISTUTILS-AND-WINDOWS.py`, but you can't import dashed module names normally.

File: tools/distrib/python/grpcio_tools/setup.py (hunk 7 above)
Comment (soltanmm) [inline tools/distrib/python/grpcio_tools/setup.py]: If you mean whether or not the flag needs to be defined somewhere: yes, it's needed.
If you mean, "Isn't this already in the build scripts?" then: this is 'needed' because I want source distributions to eventually be self-contained and buildable on our target systems without requiring that users have intricate knowledge of their compilers and environments in the common case, and I want the flags in the build scripts to eventually be automagical.

File: src/python/grpcio/build.py (hunk 3 above)
Comment (soltanmm) [inline src/python/grpcio/build.py]: Hm. My initial reaction was, "I have no guarantee that `spawn` isn't going to be used in other cross-cutting ways by future `distutils` code in later Py3 revisions, while I have a pretty strong semver guarantee that I'm safe until Py4 since I'm maintaining the semantics that're now present in Py3 for this function."

Maybe code simplicity and maintenance comfort is preferable, though. While I'm not quite yet on the fence about it, I'm leaning rather heavily on a weak board and I can hear some rusted nails popping.

File: tools/run_tests/build_python.sh (hunk 5 above)
Comment (kpayson64) [inline tools/run_tests/build_python.sh]: Is there more work involved in this than I'm assuming?  If its a couple lines of code I would rather it be done as part of this PR.

File: src/python/grpcio/build.py (hunk 3 above)
Comment (kpayson64) [inline src/python/grpcio/build.py]: CCompiler Spawn is a documented part of the distuitils API, so its use shouldn't change in the Py3 line https://docs.python.org/3/distutils/apiref.html

I would argue that the bug is in ```spawn()``` to begin with, the issue in ```link()``` is just a symptom because it happens to generate a large enough command for spawn.  I think we should just pipe all our commands to a file, and then use the file, and then use the command file hack for everything.

File: tools/distrib/python/grpcio_tools/setup.py (hunk 7 above)
Comment (kpayson64) [inline tools/distrib/python/grpcio_tools/setup.py]: Makes sense.

File: PYTHON-MANIFEST.in (hunk 6 above)
Comment (kpayson64) [inline PYTHON-MANIFEST.in]: _unixccompiler_patch.py perhaps?

File: src/python/grpcio/build.py (hunk 3 above)
Comment (soltanmm) [inline src/python/grpcio/build.py]: I *think* I see the discrepancy in understanding.

As I understand it, command files are a `gcc` specific feature. They're not a feature of the shell. Shell command files (e.g. `.bat`, `.sh`) will *also* have the character limit.

That doesn't bar special-casing `spawn()` to dissect the command, identify if the head of the command is a string corresponding to a utility that accepts the command file argument, and then proceeding from there.

File: PYTHON-MANIFEST.in (hunk 6 above)
Comment (soltanmm) [inline PYTHON-MANIFEST.in]: Sure.

File: tools/run_tests/build_python.sh (hunk 5 above)
Comment (soltanmm) [inline tools/run_tests/build_python.sh]: I could do it in a couple of lines of code, but that'd break the spirit of trying to [clean up the build files](https://github.com/grpc/grpc/pull/7195#discussion_r69870213). I think what I'd rather do is just address the TODO on line 130 and make this TODO on 139 irrelevant. Which can happen in this PR.

File: src/python/grpcio/build.py (hunk 3 above)
Comment (kpayson64) [inline src/python/grpcio/build.py]: On linux, they might be a gcc specific feature.  On Windows however, it looks like command files can be used for pretty much any command. https://support.microsoft.com/en-us/kb/830473

If we case this patch on windows, we shouldn't need to do any special casing in ```spawn()``` to dissect the command.

Sorry for the back and forth, but if we copy-and-paste 100 lines of distuitils into our code base, I just want to be sure that all other options have been explored.

File: tools/run_tests/build_python.sh (hunk 5 above)
Comment (soltanmm) [inline tools/run_tests/build_python.sh]: Hm. The change grew a little quickly. It's probably large enough to deserve its own PR. Mind waiting on the flag movement until post-merge?

Comment (kpayson64) [inline tools/run_tests/build_python.sh]: That's fine

File: src/python/grpcio/build.py (hunk 3 above)
Comment (soltanmm) [inline src/python/grpcio/build.py]: Are you sure you read that article all the way through? Command-files are an opt-in feature only stylistically recommended. Further, `gcc` accepts command files in a different manner than the toy example provided by Microsoft.

Comment (soltanmm) [inline src/python/grpcio/build.py]: How 'bout I slap a `TODO` on there to replace the hack with a less ugly hack (e.g. dissecting the command for `spawn`) and we call it an unblocking-windows-testing-day?

Comment (kpayson64) [inline src/python/grpcio/build.py]: I'm still not convinced we should go with this, but I'll defer,

@nathanielmanistaatgoogle 
Do you have any thoughts on this?

Comment (kpayson64) [inline src/python/grpcio/build.py]: Here's what I have in mind

```
def _spawn_nt(cmd):
  temporary_directory = tempfile.mkdtemp()
  command_filename = os.path.abspath(
    os.path.join(temporary_directory, 'command'))
  with open(command_filename, 'w') as command_file:
    escaped_args = [arg.replace('\\', '\\\\') for arg in cmd[1:]]
    command_file.write(' '.join(escaped_args))
  disutils.spawn.spawn([cmd[0], command_filename])

def monkeypatch_unix_compiler():
  if os.name == 'nt':
    nixccompiler.UnixCCompiler.spawn = _spawn_nt

Comment (soltanmm) [inline src/python/grpcio/build.py]: Fair enough, but given that command files are non-standardized, it'd still need to dissect the command and check that it's running with e.g. `gcc`, or `clang`, or `ccache XYZ` (wherein there'd need to be a `@` prefix on the command file).

I feel like we've gone beyond the scope of the PR. I'm not against doing something like that in the future, and have included a TODO. That said, the Windows tests are a GA blocker; trading one monkeypatch for another monkeypatch is not.

Comment (kpayson64) [inline src/python/grpcio/build.py]: From what I can tell, windows command files are standardized and don't require the @ prefix because they can be handled by arbitrary executables, and are not specific to gcc/clang/ccache.  If this patch only gets applied to Windows, we shouldn't need to dissect the command at all.

I feel pretty strongly against copying in chunks of distutils to our code base.  I'd like to hear at least one other person say this is a reasonable approach for now.

Comment (soltanmm) [inline src/python/grpcio/build.py]: Okay - look. I don't mind waiting for another person's input, but I have to at least ensure that we reach a consensus on the model of the situation. The following may include contrived examples, but work with me here:

1. Running `gcc.exe $FILENAME_HERE` is going to result in `gcc` trying to compile the file in both PowerShell and MSYS2 `bash`.
2. Running `notepad.exe $FILENAME_HERE` is going to result in `notepad.exe` trying to open the file.
3. Running `gcc @$FILENAME_HERE` is going to result in `gcc` interpreting `$FILENAME_HERE` as a command file on MSYS2 `bash`, and PowerShell getting confused about the splatting operator.
4. Running `notepad.exe @$FILENAME_HERE` is going to result in PowerShell wondering why you want to use the splatting operator, and in `bash` it'll try to open the literal file `"@$FILENAME_HERE"`.

Where are you getting that this is standardized, and even if it is how does that connect to our expected use of `gcc` (and `clang`, and maybe eventually `ccache gcc` or `ccache clang`) on MSYS2 `bash`?

Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/build.py]: @kpayson64: I don't have any particular thoughts here and I'm fairly certain that the two of you are both more familiar with the details than I am.

**Comment by kpayson64** [comment]
LGTM after rebase

**Comment by soltanmm** [comment]
Still kicking the tests...

**Comment by soltanmm** [comment]
Portability test results are consistently insane; this change did not touch C++ code in any way that ought to be running for C++-specific builds, and yet... failures when building the `protobuf` library (not `grpcio-tools`) abount.

**Comment by soltanmm** [comment]
Something got screwed up with the submodules. https://grpc-testing.appspot.com/job/gRPC_pull_requests/10711/config=opt,language=sanity,platform=linux/testReport/junit/(root)/tests/tools_run_tests_sanity_check_submodules_sh/

But the `.gitmodules` file was never touched. Hm.

**Comment by soltanmm** [comment]
Hah, turns out I was wrong. Somewhere in there the wrong version of `protobuf` got checked out and committed... y'know, I still haven't the darnedest idea how that happened.

File: PYTHON-MANIFEST.in (hunk 6 above)
Comment (kpayson64) [inline PYTHON-MANIFEST.in]: I think this line still needs to be changed to _unixccompiler_patch.py

**Comment by kpayson64** [comment]
test this please

**Comment by kpayson64** [comment]
Relevant tests passed, marking ready to merge.

File: PYTHON-MANIFEST.in (hunk 6 above)
Comment (kpayson64) [inline PYTHON-MANIFEST.in]: Forgot to follow up on this in my merge frenzy...

Comment (soltanmm) [inline PYTHON-MANIFEST.in]: Derp. Oh well, we have it now, and we'll have source-distrib tests soon-ish.
